synse/websocket.go

```diff
@@ -0,0 +1,515 @@
+// later be used by an atomic function, which makes it more concurrency-safe.
+// FIXME - is it necessary though?
+var counter uint64
+
+type websocketClient struct {
+	// options is the global config options of the client.
+	options *Options
+
+	// client holds the websocket.Dialer.
+	client *websocket.Dialer
+
+	// connection holds the websocket connection.
+	connection *websocket.Conn
+
+	// apiVersion is the current api version of Synse Server that we are
+	// communicating with.
+	apiVersion string
+
+	// entryRoute is the entry route to start the websocket connection.
+	entryRoute string
+
+	// scheme could either be ws or wss depending on TLS configuration.
+	scheme string
+}
+
+// NewWebSocketClientV3 returns a new instance of a websocket client for v3.
+func NewWebSocketClientV3(opts *Options) (Client, error) {
+	c, err := createWebSocketClient(opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create a websocket client")
+	}
+
+	s := "ws"
+	if opts.TLS.Enabled {
+		s = "wss"
+	}
+
+	return &websocketClient{
+		options:    opts,
+		client:     c,
+		apiVersion: "v3",
+		entryRoute: "connect",
+		scheme:     s,
+	}, nil
+}
+
+// createWebSocketClient setups a websocket dialer with configured options.
+func createWebSocketClient(opts *Options) (*websocket.Dialer, error) {
+	err := setDefaults(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if !opts.TLS.Enabled {
+		return &websocket.Dialer{
+			HandshakeTimeout: opts.WebSocket.HandshakeTimeout,
+		}, nil
+	}
+
+	// Only return a TLS client if its config option is enable.
+	cert, err := setTLS(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &websocket.Dialer{
+		HandshakeTimeout: opts.WebSocket.HandshakeTimeout,
+		TLSClientConfig: &tls.Config{
+			Certificates: []tls.Certificate{cert},
+		},
+	}, nil
+}
+
+// Open opens the websocket connection between the client and Synse Server.
+func (c *websocketClient) Open() error {
+	conn, _, err := c.client.Dial(buildURL(c.scheme, c.options.Address, c.apiVersion, c.entryRoute), nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to open the websocket connection")
+	}
+
+	c.connection = conn
+	return nil
+}
+
+// Close closes the websocket connection between the client and Synse Server.
+// It's up to the user to close the connection after finish using it.
+// FIXME - it makes sense to use defer function with Close. However, since
+// Close returns an error, do user need to check it manually? If yes, how?
+func (c *websocketClient) Close() error {
+	err := c.connection.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+	if err != nil {
+		return errors.Wrap(err, "failed to close the connection gracefully")
+	}
+
+	return nil
+}
+
+// Status returns the status info. This is used to check if the server
+// is responsive and reachable.
+// NOTE - this method is not applicable for websocket client.
```

Comment:
what we could do is either ping the connection, if that is supported and return an OK response here if that succeeds, or just return OK if the connection was set up properly.